// jsreport-studio/src/components/Modals/CloseConfirmationModal.js
import React, {Component, PropTypes} from 'react'
import {connect} from 'react-redux'
import {selectors} from '../../redux/entities'
import {actions} from '../../redux/editor'

@connect((state, props) => ({ entity: selectors.getById(state, props.options._id) }), { ...actions })
export default class CloseConfirmationModal extends Component {
  static propTypes = {
    close: PropTypes.func.isRequired,
    options: PropTypes.object.isRequired
  }

  remove () {
    this.props.closeTab(this.props.entity._id)
    this.props.close()
  }

  cancel () {
    this.props.close()
  }

  componentDidMount () {
    setTimeout(() => this.refs.cancel.focus(), 0)
  }

  render () {
    const { entity } = this.props
    return <div>
      <div>Are you sure you want to close {entity.name} and lose the changes ? </div>

      <div className='button-bar'>
        <button className='button danger' onClick={() => this.remove()}>Yes</button>
        <button className='button confirmation' ref='cancel' onClick={() => this.cancel()}>Cancel</button>
      </div>
    </div>
  }
}
