// jsreport-studio/lib/requestLog.js
module.exports = function (reporter) {
  reporter.afterRenderListeners.add('requestsLog', this, function (request) {
    return reporter.settings.findValue('requestsLog').then(function (v) {
      v = v || []
      v.unshift({
        template: { shortid: request.template.shortid },
        timestamp: new Date(),
        logs: request.logs || []
      })

      if (v.length > 5) {
        v.pop()
      }

      return reporter.settings.addOrSet('requestsLog', v)
    })
  })

  if (!reporter.renderErrorListeners) {
    return
  }

  reporter.renderErrorListeners.add('failedRequestsLog', this, function (request, response, err) {
    return reporter.settings.findValue('failedRequestsLog').then(function (v) {
      v = v || []
      v.unshift({
        template: { shortid: request.template.shortid },
        timestamp: new Date(),
        logs: request.logs || [],
        error: {
          message: err.message,
          stack: err.stack
        }
      })

      if (v.length > 5) {
        v.pop()
      }

      return reporter.settings.addOrSet('failedRequestsLog', v)
    })
  })
}
